integrations-and-supported-tools/pytorch/scripts/Neptune_Pytorch_Support_Advanced.py

```diff
@@ -0,0 +1,182 @@
+
+    epoch_loss = 0.0
+    epoch_acc = 0.0
+    best_acc = 0.0
+
+    # Step 3: Log losses and metrics 
+    for epoch in range(params['epochs']):
+        running_loss = 0.0
+        running_corrects = 0
+        
+        for i, (x, y) in enumerate(trainloader, 0):
+            x, y = x.to(params['device']), y.to(params['device'])
+            optimizer.zero_grad()
+            outputs = model.forward(x)
+            _, preds = torch.max(outputs, 1)
+            loss = criterion(outputs, y)
+            acc = (torch.sum(preds == y.data)) / len(x)
+
+            # Log batch loss
+            run["training/batch/loss"].log(value = loss)
+
+            # Log batch accuracy
+            run["training/batch/acc"].log(value = acc)
+
+            loss.backward()
+            optimizer.step()
+        
+            running_loss += loss.item()
+            running_corrects += torch.sum(preds == y.data)
+            
+        epoch_loss = running_loss/dataset_size['train']
+        epoch_acc = running_corrects.double().item() / dataset_size['train']
+
+        # Log epoch loss
+        run[f"training/epoch/loss"].log(value = epoch_loss, step = epoch + 1)
+
+        # Log epoch accuracy
+        run[f"training/epoch/acc"].log(value = epoch_acc, step = epoch + 1)
+
+        print(f'Epoch:{epoch+1}, Loss: {epoch_loss}, Acc: {epoch_acc}')
+        if epoch_acc > best_acc:
+            best_acc = epoch_acc
+
+            # Saving model arch & weights
+            save_model(model, params['model_filename'])
+            print('Saving model -- Done!')
+
+
+    # Step 4: Log model arch & weights -- > link to adding artifacts
+    run[f"io_files/artifacts/{params['model_filename']}_arch"].upload(f"./{params['model_filename']}_arch.txt")
+    run[f"io_files/artifacts/{params['model_filename']}"].upload(f"./{params['model_filename']}.pth")
+
+
+    # Step 5: Log Images & Predictions
+    classes = ['airplane', 'automobile', 'bird', 'cat', 'deer', 'dog', 'frog','horse','ship','truck']
+    dataiter = iter(validloader)
+    images, labels = dataiter.next()
+    images_arr = []
+    labels_arr = []
+    pred_arr = []
+    model.eval()
+    # moving model to cpu for inference 
+    if torch.cuda.is_available(): model.to("cpu")
+    # iterating on the dataset to predict the output
+    for i in range(0,10):
+        images_arr.append(images[i].unsqueeze(0))
+        labels_arr.append(labels[i].item())
+        ps = torch.exp(model(images_arr[i]))
+        ps = ps.data.numpy().squeeze()
+        pred_arr.append(np.argmax(ps))
+    # plotting the results
+    fig = plt.figure(figsize=(25,4))
+    for i in range(10):
+        ax = fig.add_subplot(2, 20/2, i+1, xticks=[], yticks=[])
+        ax.imshow(images_arr[i].squeeze().permute(2,1,0).numpy().clip(0,1))
+        ax.set_title("{} ({})".format(classes[pred_arr[i]], classes[labels_arr[i]]),
+                    color=("green" if pred_arr[i]==labels_arr[i] else "red"))
+
+    fig.savefig('predictions.jpg')
```

Comment:
you can log it directly from .jpg

Also the support neptune has for pytorch is actually in you not having to do all this boilerplate
You can log pytorch tensors directly to neptune

https://docs.neptune.ai/you-should-know/what-can-you-log-and-display#pytorch-tensor

Reply:
Thanks, this really helped I reduced 5-6 lines of code and replaced it with 1